### PR TITLE
Feat/post likes

### PR DESCRIPTION
**Pull request by karljerico**
### OVERVIEW

- Initial UI for like and unlike button
- JavaScript for like and unlike button that dont reload the page
- Added model for post likes and associate it to the user and post model
- Added table for post likes
- Added controller for post likes

### SNIPPETS
image
image
image
image
image
image

### CODE REVIEW

File: src/app/Http/Controllers/LikeController.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+use Illuminate\Http\JsonResponse;
+
+class LikeController extends Controller
+{
+    public function like(Post $post): JsonResponse
+    {
+        $user = auth()->user();
+
+        if ($user) {
+            if (!$user->likes->contains('post_id', $post->id)) {
```
Comment (bpocrafael) [inline src/app/Http/Controllers/LikeController.php]: too much nested.

File: src/app/Http/Controllers/LikeController.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        return response()->json();
+    }
+
+    public function unlike(Post $post): JsonResponse
+    {
+        $user = auth()->user();
+
+        if ($user) {
+            $like = $user->likes()->where('post_id', $post->id)->first();
+
```
Comment (bpocrafael) [inline src/app/Http/Controllers/LikeController.php]: too much nested

Comment (bpocrafael) [inline src/app/Http/Controllers/LikeController.php]: add service layer for this one

File: src/app/Models/Post.php [hunk 3]
```diff
@@ -18,9 +20,30 @@ class Post extends Model
...
+     */
+    public function likes(): HasMany
+    {
+        return $this->hasMany(PostLike::class);
+    }
+
+    /**
+     * Check if the post is liked by the given user.
+     *
+     * @param User $user
```
Comment (bpocrafael) [inline src/app/Models/Post.php]: remove php docs

File: src/app/Models/PostLike.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    protected $fillable = [
+        'user_id',
+        'post_id',
+    ];
+
+    /**
+    * Get the user associated with the like.
+    *
+    * @return \Illuminate\Database\Eloquent\Relations\BelongsTo
```
Comment (bpocrafael) [inline src/app/Models/PostLike.php]: remove php docs

File: src/public/js/likeToggle.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
+    document.addEventListener('DOMContentLoaded', function () {
+    const likeForms = document.querySelectorAll('.like-form');
+    const unlikeForms = document.querySelectorAll('.unlike-form');
+
+    likeForms.forEach(function (likeForm) {
+        likeForm.addEventListener('submit', function (e) {
+            e.preventDefault(); // Prevent the default form submission
```
Comment (bpocrafael) [inline src/public/js/likeToggle.js]: remove comments

File: src/public/js/likeToggle.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+            likeForm.querySelector('.like').classList.add('d-none');
+            likeForm.nextElementSibling.querySelector('.unlike').classList.remove('d-none');
+
+            fetch(likeForm.action, {
+                method: 'POST',
+                headers: {
+                    'X-CSRF-TOKEN': likeForm.querySelector('input[name="_token"]').value,
```
Comment (bpocrafael) [inline src/public/js/likeToggle.js]: ???

Comment (karljerico) [reply src/public/js/likeToggle.js:14]: Goodafternoon po sir, any context po for this comment? Thank you po. Hindi ko po alam if ano po gagawain 🙇🙇

File: src/public/js/likeToggle.js (hunk 5 above)
Comment (nadj1233) [reply src/public/js/likeToggle.js]: why do you loop the form?

File: src/app/Http/Controllers/LikeController.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+class LikeController extends Controller
+{
+    protected LikeService $likeService;
+
+    public function __construct(LikeService $likeService)
+    {
+        $this->likeService = $likeService;
+    }
+
+    public function like(Request $request): JsonResponse
```
Comment (hipe01mark) [inline src/app/Http/Controllers/LikeController.php:21]: add comment

File: src/app/Services/LikeService.php [hunk 8]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace App\Services;
+
+use App\Models\PostLike;
+
+class LikeService
+{
+    public function like(int $user_id, int $post_id): bool
```
Comment (hipe01mark) [inline src/app/Services/LikeService.php:12]: add comment

File: src/app/Services/LikeService.php [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        $like = PostLike::where('user_id', $user_id)
+            ->where('post_id', $post_id)
+            ->first();
+
+        if ($like) {
+            $like->delete();
+        } else {
```
Comment (hipe01mark) [inline src/app/Services/LikeService.php]: Prevent using else in the if statement.

File: src/public/js/likeToggle.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+            type: 'POST',
+            url: `/posts/${post_id}/like`,
+            data: {
+                _token: csrfToken,
+                post_id: post_id
+            },
+            success: function(response) {
+                if (response.liked) {
+                    button.text('Unlike');
+                } else {
```
Comment (hipe01mark) [inline src/public/js/likeToggle.js]: prevent using else in if statement

**Review by hipe01mark — APPROVED** [review]
LGTM :bow:

File: src/public/js/likeToggle.js (hunk 5 above)
Comment (karljerico) [reply src/public/js/likeToggle.js]: Comment is resolved, restructured the js with AJAX. Thank you 🙇🙇

**Review by hipe01mark — APPROVED** [review]
LGTM 🙇

**Review by nadj1233 — APPROVED** [review]
LGTM 💯

**Pull request merged**